proposals/20240606-jwt-handling.md

```diff
@@ -0,0 +1,135 @@
+    1. if the mesh will not override these headers, then users could forge requests and impersonate other users by setting the header and any valid token
+    1. for User to Machine traffic, the `email` claim from `id_token` should be used by default but it should also allow parameterization to allow using different claims. This is doable via `RequestAuthentication` with object per issuer
+    2. in case of K8s ServiceAccount tokens the `sub` claim will be used. `sub` claim format is `system:serviceaccount:<sa-namespace>:<sa-name>`
+    3. in case of Dex tokens it would be `kimonas@email.com`
+5. The backends will use the information from the headers and not deal with JWTs
+    1. `SubjectAccessReviews` should be made for the `user` and `groups` that were exposed from the headers, independently of the issuer of the token
+    2. [`SubjectAccessReview API`](https://kubernetes.io/docs/reference/kubernetes-api/authorization-resources/subject-access-review-v1/) uses the `user`,
+       `groups`, `resource` and `verb` details to verify the access against K8s RBAC, which allows defining authorization to specific actions based on
+       K8s standard RBAC implementation
+
+With the above implementation we move all the logic of handling the JWTs to the service-mesh and leave only the business logic to the apps. This also means that the apps don't care about token "types" (i.e. dex, k8s tokens etc) and only have to look at the corresponding headers.
+
+This proposal aims to put more focus on keeping and validating `id_tokens` but also bridging to the existing functionality of the backends, to avoid extensive changes.
+### Implementation
+The technical details for the above proposal translate to the following
+1. Common Kubeflow manifests, for all components, for configuring Istio for supporting multiple issuers ([Dex](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/oidc-client/oauth2-proxy/components/istio-external-auth/requestauthentication.dex-jwt.yaml) and [K8s-m2m](https://github.com/kubeflow/manifests/blob/v1.9-branch/common/oidc-client/oauth2-proxy/components/istio-m2m/requestauthentication.yaml)), via `RequestAuthentication` objects
+2. `AuthorizationPolicy` objects of components, for allowing access from Istio IngressGateway, will need to be extended for also requiring a JWT
+3. Backends that need to be accessible from other user-namespaces will need to have an `AuthorizationPolicy` that allows any request, only if it has a JWT
+4. Backends don't need any logic for validating the JWTs and their existence
+6. `RequestAuthentication` objects, per issuer, should expose the corresponding token claims to the `kubeflow-userid` and `kubeflow-groups` headers
+7. Backends only need to care about `kubeflow-userid` and `kubeflow-groups` headers
+#### Requiring a JWT
+The service-mesh will need to drop requests (403) that don't have any JWT, for apps that expect user identity in the requests.
```

Comment:
The service-mesh will need to drop requests (403) that don't have any JWT, for services that expect a user identity in the requests.

Reply:
pushed a commit a4ba96a4